aptos-move/framework/aptos-token/sources/token.move

```diff
@@ -1475,6 +1475,71 @@ module aptos_token::token {
         token_data.royalty
     }
 
+    /// return the token_data_id from the token_id
+    public fun get_tokendata_id(token_id: TokenId): TokenDataId {
+        token_id.token_data_id
+    }
+
+    /// return the mutation setting of the token
+    public fun get_tokendata_mutability_config(token_data_id: TokenDataId): TokenMutabilityConfig acquires Collections {
+        let creator_addr = token_data_id.creator;
+        assert!(exists<Collections>(creator_addr), error::not_found(ECOLLECTIONS_NOT_PUBLISHED));
+        let all_token_data = &borrow_global<Collections>(creator_addr).token_data;
+        assert!(table::contains(all_token_data, token_data_id), error::not_found(ETOKEN_DATA_NOT_PUBLISHED));
+        table::borrow(all_token_data, token_data_id).mutability_config
+    }
+
+    /// return if the token's maximum is mutable
+    public fun get_token_mutability_maximum(config: &TokenMutabilityConfig): bool {
+        config.maximum
+    }
```

Comment:
do we have an example use case here?
the paradigm you have chosen is to perform a copy for the config and then pass that into the getters

not saying this is right or wrong, I just want to understand if this is just out of intuition or if there's a reason why

Reply:
Here is an example of trying to iterate the config to find fields that are immutable. 
`public fun get_unmutable_fields(token_id: TokenId): vector<String> {
        let token_data_id = token::get_tokendata_id(token_id);
        let (_, _, _, property_version) = token::get_token_id_fields(&token_id);
        let mutability_config = token::get_tokendata_mutability_config(token_data_id);
        if (!token::get_token_mutability_maximum(&mutability_config)) {
             // add maximum as immutable ?  
        };
       if (!token::get_token_mutability_description(&mutability_config)) {
             // add descrption as immutable ?  
        };
    }'

Reply:
It's only 5 bools right? Why not just have a single getter function that reads the resource once and returns all 5? It wouldn't be a lot more gas overhead.

Reply:
Ok, this works as well. I don't have any preference here.

Reply:
updated to return bool tuple directly.

Reply:
Am I the only one think the original way is more organized? (btw, Can `public fun get_tokendata_mutability_config(token_data_id: TokenDataId): TokenMutabilityConfig` return `&TokenMutabilityConfig` as it comes from table)
Returning a (bool, bool, ...) definitely looks like a hack instead of standard library method.
Plus, I expect as Move in evolving, we will support returning global reference to avoid unnecessary copy and may have default getter implemented as solidity does.

Reply:
I think returning `TokenMutabilityConfig` is a bit more hassle for the sdk folks since they'll need to create the struct in Rust/TypeScript to deserialize and get the value. And in this case it's the value that actually matters?

Reply:
These are two topics.
1. An organized and explicit api
2. write sdk for api.

There are related but you are not supposed to break the granularity of an API for the convenience of 2. The hassle you mentioned can be solved by creating tools like codegen.

Reply:
Accessing via a struct is only beneficial if the struct can be involved and more fields or complex structure can be added. In this case, however, structs cannot be extended (compat reasons) and the layout here is simple enough that it's cleaner to have a single function to access the entire state rather than 2 - 1 to return the struct (which involves copying and costs extra gas) and 1 to extract each field.